fairscale/nn/pipe/ampnet.py

```diff
@@ -0,0 +1,347 @@
+                optimizer.step()
+                optimizer.zero_grad()
+
+    def event_loop_tail_across_minibatches(
+        self, lm_dataloader: DataLoader, criterion: nn.Module, optimizer: Optimizer, vocab_size: int
+    ) -> None:
+        # handles one epoch
+
+        cur_rank = self.group.rank()
+        N = len(get_pipeline_parallel_ranks())
+        num_batches = len(lm_dataloader)
+        lm_iter = enumerate(lm_dataloader)
+        # last partition -> one forward / one backward -> no warmup
+        count = 0
+        num_gradients = 0
+        activations = dict()
+
+        log_interval = 1
+        word_counter = 0
+        total_loss = 0
+
+        while True:
+            try:
+                start_time = time.time()
+                microbatch_index, cur_batch = next(lm_iter)
+                reqd_target = cur_batch["target"].to(self.input_device)
+
+                # one forward
+                message = self.transport.recv_message_header(EVENT_LOOP_ACTIVATIONS_QUEUE)
+                args: AsyncMessageBody = message.args
+                assert args.microbatch_index == count
+                batch = self.get_batch_from_message(message, EVENT_LOOP_GRADIENTS_QUEUE)
+
+                if self.weight_prediction:
+                    optimizer.update_weight_using_future_predictions(cur_rank, N, forward=True)  # type: ignore
+                task = create_task_without_skip_trackers(
+                    self.checkpoint_stop, args.microbatch_index, self.group.rank(), batch, self.partitions[0].module,
+                )
+                output = task.compute()
+                activations[args.microbatch_index] = output
+                task.finalize(output)
+                # one backward
+                if self.weight_prediction:
+                    optimizer.update_weight_using_future_predictions(cur_rank, N, forward=False)  # type: ignore
+
+                if vocab_size == 0:
+                    loss = criterion(output.tensor.view(-1), reqd_target.view(-1))
+                else:
+                    op = output.tensor.view(-1, vocab_size)
+                    rt = reqd_target.view(-1)
+                    loss = criterion(op, rt)  # for LM usecase
+
+                loss.backward()
+                count += 1
+                num_gradients += 1
+
+                if (optimizer is not None and num_gradients % self.min_update_interval == 0) or self.weight_prediction:
+                    optimizer.step()
+                    optimizer.zero_grad()
+
+                if vocab_size != 0:  # for real use-case
+                    word_counter += cur_batch["ntokens"]
+                    if count % log_interval == 0 and count > 0:
+                        total_loss += loss.item()
+                        cur_loss = total_loss / log_interval
+                        elapsed = time.time() - start_time
+                        print(
```

Comment:
Use log instead of print.

Reply:
Ohh, this was the old PR. I just created a new one: https://github.com/facebookresearch/fairscale/pull/304